File: wliotproxy-src/libwliotproxy/IotServerConnection.cpp
/*******************************************
Copyright 2017 OOO "LMS"

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

	http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.*/

#include "IotServerConnection.h"
#include "IotServerStoragesCommands.h"
#include "IotServerCommandCall.h"
#include <QDebug>
#include <QEventLoop>
#include <QCoreApplication>

IotServerConnection::IotServerConnection(QObject *parent)
	:QObject(parent)
{
	sock=0;
	callIdNum=0;
	wasSyncMsg=false;
	ready=false;
	proxy=QNetworkProxy(QNetworkProxy::NoProxy);
	syncTimer.setInterval(WLIOTProtocolDefs::syncWaitTime*2);
	syncTimer.setSingleShot(false);
	connect(&parser,&StreamParser::newMessage,this,&IotServerConnection::onRawMessage);
	connect(&syncTimer,&QTimer::timeout,this,&IotServerConnection::onSyncTimer,Qt::QueuedConnection);
}

IotServerConnection::~IotServerConnection()
{
	if(isConnected())
	{
		disconnectFromServer();
		onDevDisconnected();
	}
}

bool IotServerConnection::startConnectLocal()
{
	if(sock)return false;
	netConn=false;
	ready=false;
	parser.reset();
	sockThread.start();
	while(!sockThread.isRunning())
		QThread::yieldCurrentThread();
	sock=new IotServerConnectionSocketWrap(this);
	sock->moveToThread(&sockThread);
	QMetaObject::invokeMethod(sock,"startConnectLocal",Qt::QueuedConnection);
	return true;
}

bool IotServerConnection::startConnectNet(const QString &host,quint16 port)
{
	if(sock)return false;
	mHost=host;
	mPort=port;
	netConn=true;
	ready=false;
	parser.reset();
	sockThread.start();
	while(!sockThread.isRunning())
		QThread::yieldCurrentThread();
	sock=new IotServerConnectionSocketWrap(this);
	sock->moveToThread(&sockThread);
	QMetaObject::invokeMethod(sock,"startConnectNet",Qt::QueuedConnection);
	return true;
}

bool IotServerConnection::authenticateNet(const QByteArray &userName,const QByteArray &pass)
{
	if(!sock->netSock->isEncrypted())
		return false;
	QByteArrayList retVal;
	bool oldNetAuth=ready;
	ready=true;//if not this, call will not be processed
	if(!execCommand(WLIOTServerProtocolDefs::authenticateSrvMsg,QByteArrayList()<<userName<<pass,retVal))
		return false;
	ready=oldNetAuth;
	bool ok=false;
	uid=retVal.value(0).toLong(&ok);
	if(!ok)
		uid=-1;
	if(!ready)
	{
		ready=true;
		emit preconnected();
		qApp->processEvents(QEventLoop::ExcludeUserInputEvents);
		emit connected();
		wasSyncMsg=true;
		syncTimer.start();
	}
	emit authenticationChanged();
	return true;
}

bool IotServerConnection::authenticateLocalFromRoot(const QByteArray &userName)
{
	if(!sock->localSock->isOpen())return false;
	QByteArrayList retVal;
	if(!execCommand(WLIOTServerProtocolDefs::authenticateSrvMsg,QByteArrayList()<<userName,retVal))
		return false;
	bool ok=false;
	uid=retVal.value(0).toLong(&ok);
	if(!ok)
		uid=-1;
	emit authenticationChanged();
	return true;
}

bool IotServerConnection::isConnected()
{
	return ready;
}

bool IotServerConnection::execCommand(const QByteArray &cmd,
	const QByteArrayList &args,QByteArrayList &retVal,CmDataCallback onCmData)
{
	IotServerCommandCall call(this,onCmData,cmd,args);
	call.call();
	retVal=call.returnValue();
	return call.ok();
}

bool IotServerConnection::execCommand(const QByteArray &cmd,const QByteArrayList &args,CmDataCallback onCmData)
{
	QByteArrayList retVal;
	return execCommand(cmd,args,retVal,onCmData);
}

bool IotServerConnection::waitForConnected(int msec)
{
	if(!sock)return false;
	if(isConnected())
		return true;
	QEventLoop loop;
	QTimer t;
	t.setInterval(msec);
	t.setSingleShot(true);
	connect(this,&IotServerConnection::connected,&loop,&QEventLoop::quit);
	connect(this,&IotServerConnection::connectionError,&loop,&QEventLoop::quit);
	connect(&t,&QTimer::timeout,&loop,&QEventLoop::quit);
		loop.exec(QEventLoop::ExcludeUserInputEvents);
	return isConnected();
}

void IotServerConnection::disconnectFromServer()
{
	if(!sock)return;
	QMetaObject::invokeMethod(sock,"disconnectFromServer",Qt::QueuedConnection);
}

bool IotServerConnection::subscribeStorage(const QByteArray &devIdOrName,const QByteArray &sensorName)
{
	return execCommand("subscribe",QByteArrayList()<<devIdOrName<<sensorName);
}

bool IotServerConnection::unsubscribeStorage(const QByteArray &devIdOrName,const QByteArray &sensorName)
{
	return execCommand("unsubscribe",QByteArrayList()<<devIdOrName<<sensorName);
}

bool IotServerConnection::identifyServer(QUuid &id,QByteArray &name)
{
	QByteArrayList retVal;
	if(!execCommand(WLIOTProtocolDefs::identifyMsg,QByteArrayList(),retVal)||retVal.count()<2)
		return false;
	id=QUuid(retVal[0]);
	name=retVal[1];
	return !id.isNull();
}

bool IotServerConnection::writeMsg(const Message &m)
{
	if(!sock)return false;
	QByteArray msgData=m.dump();
	QMetaObject::invokeMethod(sock,"writeData",Qt::QueuedConnection,Q_ARG(QByteArray,msgData));
	return true;
}

void IotServerConnection::setProxy(const QNetworkProxy &p)
{
	proxy=p;
}

bool IotServerConnection::writeVDevMsg(const QUuid &id, const Message &m)
{
	return writeMsg(Message(WLIOTServerProtocolDefs::vdevMsg,QByteArrayList()<<id.toByteArray()<<m.title<<m.args));
}

IotServerApmIdType IotServerConnection::userId()
{
	return uid;
}

void IotServerConnection::onNetDeviceConnected()
{
	emit needAuthentication();
}

void IotServerConnection::onDevDisconnected()
{
	qDebug()<<"iot server disconnected";
	emit disconnected();
	sock->deleteLater();
	sockThread.quit();
	sockThread.wait(3000);
	sockThread.terminate();
	parser.reset();
	sock=0;
	ready=false;
	uid=-1;
	syncTimer.stop();
}

void IotServerConnection::onRawMessage(const Message &m)
{
	if(m.title==WLIOTProtocolDefs::devSyncMsg)
	{
		qDebug()<<"iot sync";
		wasSyncMsg=true;
		writeMsg(Message(WLIOTProtocolDefs::devSyncrMsg));
	}
	else if(m.title==WLIOTProtocolDefs::measurementMsg)
	{
		if(m.args.count()<3)return;
		QUuid devId(m.args[0]);
		QByteArray sensorName=m.args[1];
		if(devId.isNull())return;
		StorageId stId={devId,sensorName};
		QMetaObject::invokeMethod(this,"newSensorValue",Qt::QueuedConnection,Q_ARG(StorageId,stId),
			Q_ARG(QByteArrayList,m.args.mid(2)));
	}
	else if(m.title==WLIOTServerProtocolDefs::notifyDeviceIdentifiedMsg)
	{
		if(m.args.count()<2)return;
		QUuid devId(m.args[0]);
		QByteArray devName=m.args[1];
		QUuid typeId;
		if(m.args.count()>2)
			typeId=QUuid(m.args[2]);
		if(devId.isNull())return;
		emit deviceIdentified(devId,devName,typeId);
	}
	else if(m.title==WLIOTServerProtocolDefs::notifyDeviceLostMsg)
	{
		if(m.args.count()<1)return;
		QUuid devId(m.args[0]);
		if(devId.isNull())return;
		emit deviceLost(devId);
	}
	else if(m.title==WLIOTProtocolDefs::stateChangedMsg)
	{
		if(m.args.count()<1)return;
		QUuid devId(m.args[0]);
		QByteArrayList args=m.args.mid(1);
		if(args.count()%3!=0)return;
		emit deviceStateChanged(devId,args);
	}
	else if(m.title==WLIOTServerProtocolDefs::notifyStorageCreatedMsg)
	{
		IotServerStorageDescr st;
		if(!IotServerStoragesCommands::storageFromArgs(m.args,st))return;
		emit storageCreated(st);
	}
	else if(m.title==WLIOTServerProtocolDefs::notifyStorageRemovedMsg)
	{
		if(m.args.count()<2)return;
		QUuid devId(m.args[0]);
		QByteArray sensorName=m.args[1];
		if(devId.isNull())return;
		emit storageRemoved({devId,sensorName});
	}
	else if(m.title==WLIOTServerProtocolDefs::vdevMsg)
	{
		if(m.args.count()<2)return;
		QUuid id(m.args[0]);
		if(id.isNull())return;
		emit vdevMsg(id,Message(m.args[1],m.args.mid(2)));
	}
	else if(m.title==WLIOTProtocolDefs::funcAnswerOkMsg||m.title==WLIOTProtocolDefs::funcAnswerErrMsg||
		m.title==WLIOTServerProtocolDefs::srvCmdDataMsg)
	{
		emit funcCallReplyMsg(m);
	}
}

void IotServerConnection::onLocalSocketConnected()
{
	qDebug()<<"IotServerConnection::onLocalSocketConnected";
	ready=true;
	emit preconnected();
	uid=0;
	emit connected();
	emit authenticationChanged();
	wasSyncMsg=true;
	syncTimer.start();
}

void IotServerConnection::onSyncTimer()
{
	qDebug()<<"iot connection onSyncTimer";
	if(wasSyncMsg)
		wasSyncMsg=false;
	else if(isConnected())
	{
		qDebug()<<"iot server sync timeout";
		/*if(netConn)
		{
			if(netSock->isOpen())
				netSock->disconnectFromHost();
			netSock->deleteLater();
			netSock=0;
		}
		else
		{
			if(localSock->isOpen())
				localSock->disconnectFromServer();
			localSock->deleteLater();
			localSock=0;
		}*/
	}
}

void IotServerConnection::onNewData(QByteArray data)
{
	parser.pushData(data);
}

File: WLIOTAndroidApplication/libwliotproxy/src/main/java/com/alterozoom/libwliotproxy/IotServerConnection.java
package com.alterozoom.libwliotproxy;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetAddress;
import java.net.Proxy;
import java.security.SecureRandom;
import java.security.cert.X509Certificate;
import java.util.ArrayList;
import java.util.UUID;

import javax.net.ssl.HandshakeCompletedEvent;
import javax.net.ssl.HandshakeCompletedListener;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSocket;
import javax.net.ssl.SSLSocketFactory;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

//TODO rewrite network read/write with multithreading and blocking queues for read and write messages

public class IotServerConnection
{
	private SSLSocket netSock;
	private InputStream iStream;
	private OutputStream oStream;
	private Thread readThread,writeThread;
	private long callIdNum;
	private boolean netAuthentificated;
	private Proxy proxy;
	private StreamParser parser;
	private ArrayList<EventsCallback> callbacks;
	private SSLSocketFactory sockFactory;
	private boolean wasSyncMsg;
	private StreamParserCb parserCb;
	public static final int defaultPort=4083;

	private class StreamParserCb
		extends StreamParser.StreamParserEventsCb
	{
		public void onNewMessage(Message m)
		{
			if(m.title.equals(WLIOTProtocolDefs.devSyncMsg))
			{
				wasSyncMsg=true;
				writeMsg(new Message(WLIOTProtocolDefs.devSyncrMsg));
			}
			else if(m.title.equals(WLIOTProtocolDefs.measurementMsg))
			{
				if(m.args.size()<3)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				ByteArray sensorName=m.args.get(1);
				if(devId==null)return;
				StorageId stId=new StorageId(devId,sensorName);
				ByteArrayList args=new ByteArrayList(m.args);
				args.remove(0);
				args.remove(0);
				for(EventsCallback cb:callbacks)
					cb.newSensorValue(stId,args);
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.notifyDeviceIdentifiedMsg))
			{
				if(m.args.size()<2)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				ByteArray devName=m.args.get(1);
				UUID typeId=null;
				if(m.args.size()>2)
					typeId=UUID.fromString(m.args.get(2).toString());
				if(devId==null)return;
				for(EventsCallback cb:callbacks)
					cb.deviceIdentified(devId,devName,typeId);
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.notifyDeviceLostMsg))
			{
				if(m.args.size()<1)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				if(devId==null)return;
				for(EventsCallback cb:callbacks)
					cb.deviceLost(devId);
			}
			else if(m.title.equals(WLIOTProtocolDefs.stateChangedMsg))
			{
				if(m.args.size()<1)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				ByteArrayList args=new ByteArrayList(m.args);
				args.remove(0);
				if(args.size()%3!=0)return;
				for(EventsCallback cb:callbacks)
					cb.deviceStateChanged(devId,args);
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.notifyStorageCreatedMsg))
			{
				IotServerStorageDescr st=new IotServerStorageDescr();
				if(!IotServerStoragesCommands.storageFromArgs(m.args,st))return;
				for(EventsCallback cb:callbacks)
					cb.storageCreated(st);
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.notifyStorageRemovedMsg))
			{
				if(m.args.size()<2)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				ByteArray sensorName=m.args.get(1);
				if(devId==null)return;
				StorageId stId=new StorageId(devId,sensorName);
				for(EventsCallback cb:callbacks)
					cb.storageRemoved(stId);
			}
			else if(m.title.equals(WLIOTServerProtocolDefs.vdevMsg))
			{
				if(m.args.size()<2)return;
				UUID devId=UUID.fromString(m.args.get(0).toString());
				if(devId==null)return;
				ByteArray devMsg=m.args.get(1);
				ByteArrayList args=new ByteArrayList(m.args);
				args.remove(0);
				args.remove(0);
				for(EventsCallback cb:callbacks)
					cb.vdevMsg(devId,new Message(devMsg,args));
			}
			else if(m.title.equals(WLIOTProtocolDefs.funcAnswerOkMsg)||
				m.title.equals(WLIOTProtocolDefs.funcAnswerErrMsg)||
				m.title.equals(WLIOTServerProtocolDefs.srvCmdDataMsg))
			{
				for(EventsCallback cb:callbacks)
					cb.funcCallReplyMsg(m);
			}
		}
	}

	public static class EventsCallback
	{
		void onPreconnected(){}
		void onConnected(){}
		void onDisconnected(){}
		void needAuthentification(){}
		void connectionError(){}
		void newSensorValue(StorageId id,ByteArrayList valueArgs){}
		void deviceIdentified(UUID id,ByteArray name,UUID typeId){}
		void deviceLost(UUID id){}
		void deviceStateChanged(UUID id,ByteArrayList args){}
		void storageCreated(IotServerStorageDescr s){}
		void storageRemoved(StorageId id){}
		void funcCallReplyMsg(Message m){}
		void vdevMsg(UUID id,Message m){}
	}

	public IotServerConnection()
	{
		callIdNum=0;
		netAuthentificated=false;
		parser=new StreamParser();
		parserCb=new StreamParserCb();
		parser.addCb(parserCb);
		try
		{
			TrustManager[] victimizedManager = new TrustManager[]
			{
				new X509TrustManager()
				{
					public X509Certificate[] getAcceptedIssuers()
					{

						X509Certificate[] myTrustedAnchors=new X509Certificate[0];
						return myTrustedAnchors;
					}

					@Override
					public void checkClientTrusted(X509Certificate[] certs, String authType)
					{
					}

					@Override
					public void checkServerTrusted(X509Certificate[] certs, String authType)
					{
					}
				}
			};

			SSLContext sc=SSLContext.getInstance("SSL");
			sc.init(null,victimizedManager,new SecureRandom());
			sockFactory=sc.getSocketFactory();
		}
		catch (Exception e)
		{
			e.printStackTrace();
		}
	}

	public void finalize()
	{
		disconnectFromServer();
		parser.removeCb(parserCb);
	}

	public void registerCallback(EventsCallback cb)
	{
		if(callbacks.contains(cb))return;
		callbacks.add(cb);
	}

	public void removeCallback(EventsCallback cb)
	{
		callbacks.remove(cb);
	}

	public boolean startConnectNet(String host,int port)
	{
		if(netSock!=null)return false;
		netAuthentificated=false;
		parser.reset();
		try
		{
			netSock=(SSLSocket)sockFactory.createSocket(InetAddress.getByName(host),port);
			netSock.startHandshake();
			netSock.addHandshakeCompletedListener(new HandshakeCompletedListener(){
				@Override
				public void handshakeCompleted(HandshakeCompletedEvent handshakeCompletedEvent)
				{
					netSock.notify();
				}
			});
			try
			{
				netSock.wait();
			}
			catch(InterruptedException e)
			{
				e.printStackTrace();
			}
			if(!netSock.isConnected())
			{
				netSock=null;
				return false;
			}
			iStream=netSock.getInputStream();
			oStream=netSock.getOutputStream();
			readThread=new Thread(new Runnable(){
				@Override
				public void run()
				{
					try
					{
						int c=iStream.read();
						if(c>0)
							parser.putByte((byte)c);
					}
					catch(IOException e)
					{
						e.printStackTrace();
					}
				}
			});
			readThread.start();
		}
		catch(IOException e)
		{
			e.printStackTrace();
			try
			{
				if(netSock!=null)
				{
					netSock.close();
					netSock=null;
				}
			}
			catch(IOException e1)
			{
				e1.printStackTrace();
			}
			return false;
		}
		return true;
	}

	public boolean authenticateNet(ByteArray userName,ByteArray pass)
	{
		if(!isConnected())
			return false;
		ByteArrayList args=new ByteArrayList();
		args.add(userName);
		args.add(pass);
		if(!execCommand(WLIOTServerProtocolDefs.authenticateSrvMsg,
			new ByteArrayListBuilder().add(userName).add(pass).done()))return false;
		netAuthentificated=true;
		for(EventsCallback cb:callbacks)
			cb.onPreconnected();
		for(EventsCallback cb:callbacks)
			cb.onConnected();
		return true;

	}

	public boolean isConnected()
	{
		if(netSock==null)
			return false;
		return netSock.isConnected();
	}

	public boolean execCommand(
		final ByteArray cmd,final ByteArrayList args,ObjectRef<ByteArrayList> retVal,CmDataCallback onCmData)
	{
		IotServerCommandCall call=new IotServerCommandCall(this,onCmData,cmd,args);
		if(retVal!=null)
			retVal.set(call.returnValue());
		return call.ok();
	}

	public boolean execCommand(final ByteArray cmd,final ByteArrayList args,CmDataCallback onCmData)
	{
		return execCommand(cmd,args,null,onCmData);
	}

	public boolean execCommand(final ByteArray cmd,final ByteArrayList args,ObjectRef<ByteArrayList> retVal)
	{
		return execCommand(cmd,args,retVal,null);
	}

	public boolean execCommand(final ByteArray cmd,final ByteArrayList args)
	{
		return execCommand(cmd,args,null,null);
	}

	public boolean execCommand(final ByteArray cmd)
	{
		return execCommand(cmd,null,null,null);
	}

	public boolean waitForConnected(int msec)
	{
		if(netSock==null)
			return false;
		return netSock.isConnected();
	}

	public void disconnectFromServer()
	{
		if(netSock==null)return;
		try
		{
			netSock.close();
		}
		catch(IOException e)
		{
			e.printStackTrace();
		}
		readThread.interrupt();
		netSock=null;
		iStream=null;
		oStream=null;
		readThread=null;
	}

	public boolean subscribeStorage(final ByteArray devIdOrName,final ByteArray sensorName)
	{
		return execCommand(ByteArray.fromString("subscribe"),
			new ByteArrayListBuilder().add(devIdOrName).add(sensorName).done());
	}

	public boolean unsubscribeStorage(final ByteArray devIdOrName,final ByteArray sensorName)
	{
		return execCommand(ByteArray.fromString("unsubscribe"),
			new ByteArrayListBuilder().add(devIdOrName).add(sensorName).done());
	}

	public boolean identifyServer(ObjectRef<UUID> id,ObjectRef<ByteArray> name)
	{
		ObjectRef<ByteArrayList> retVal=new ObjectRef<>();
		if(!execCommand(WLIOTProtocolDefs.identifyMsg,new ByteArrayList(),retVal)||retVal.o().size()<2)
			return false;
		id.set(UUID.fromString(retVal.o().get(0).toString()));
		name.set(retVal.o().get(1));
		return id!=null;
	}

	public boolean writeMsg(Message m)
	{
		if(netSock==null)
			return false;
		byte[] msgData=m.dump();
		try
		{
			oStream.write(msgData);
			oStream.flush();
		}
		catch(IOException e)
		{
			e.printStackTrace();
			return false;
		}
		return true;
	}

	public void setProxy(Proxy p)
	{
	}

	public boolean writeVDevMsg(UUID id,Message m)
	{
		return writeMsg(new Message(WLIOTServerProtocolDefs.vdevMsg,
			new ByteArrayListBuilder().add(ByteArray.fromString(id.toString())).add(m.title).add(m.args).done()));
	}
}
